From: Frank van der Linden <fllinden@amazon.com>
Date: Thu, 6 Jan 2022 19:16:23 +0000
Subject: virtio: add hack to allow pre-mapped scatterlists

When reporting offlined pages through free page reporting, and
memmap_on_memory is active, we don't want to touch the page structures
anymore, since that will lead to a reference to the range we just
offlined, as the page structures themselves reside in the range.

So, we can't use sg_phys to set the dma address. Instead, if sg_page
is set to NULL, assume that sg_dma_address is set already, and use it.

Signed-off-by: Frank van der Linden <fllinden@amazon.com>
---
 drivers/virtio/virtio_ring.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/drivers/virtio/virtio_ring.c b/drivers/virtio/virtio_ring.c
--- a/drivers/virtio/virtio_ring.c
+++ b/drivers/virtio/virtio_ring.c
@@ -360,7 +360,8 @@ static dma_addr_t vring_map_one_sg(const struct vring_virtqueue *vq,
 		 * depending on the direction.
 		 */
 		kmsan_handle_dma(sg_page(sg), sg->offset, sg->length, direction);
-		return (dma_addr_t)sg_phys(sg);
+		return sg_page(sg) == NULL ? sg_dma_address(sg) :
+		    (dma_addr_t)sg_phys(sg);
 	}
 
 	/*

